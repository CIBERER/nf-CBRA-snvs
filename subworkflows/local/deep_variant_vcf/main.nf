include { DEEPVARIANT                                                    }      from '../../../subworkflows/nf-core/deepvariant/main'
include { BCFTOOLS_FILTER                                                }      from '../../../modules/nf-core/bcftools/filter/main'


workflow DEEP_VARIANT_VCF {

    take:
    ch_bam          // channel (mandatory): [ val(meta), path(bam), path(bai) ]
    ch_intervals    // channel (mandatory) : [ val(meta), path(bed) ]
    ch_fasta        // channel (mandatory) : [ val(meta2), path(fasta) ]
    ch_fai          // channel (mandatory) : [ val(meta2), path(fai) ]

    main:

    ch_versions = Channel.empty()


    ch_input = ch_bam
    .join(ch_intervals, by: 0)
    .map { meta, bam, bai, intervals -> 
        if (intervals instanceof List && intervals.isEmpty()) {
            return [ meta, bam, bai, [] ]
        } else {
            return [ meta, bam, bai, intervals ]
        }
    }


    DEEPVARIANT (
        ch_input,
        ch_fasta,
        ch_fai,
        Channel.empty(),
        Channel.empty()
    )
    ch_versions = ch_versions.mix(DEEPVARIANT.out.versions.first())

    BCFTOOLS_FILTER(
        DEEPVARIANT.out.vcf.join(DEEPVARIANT.out.vcf_index)
    )
    ch_versions = ch_versions.mix(BCFTOOLS_FILTER.out.versions.first())
 
    vcf = BCFTOOLS_FILTER.out.vcf.join(BCFTOOLS_FILTER.out.tbi)

    emit:
    vcf                                        // channel: [ val(meta), path(vcf), path(tbi)]
    versions = ch_versions                     // channel: [ versions.yml ]

}
