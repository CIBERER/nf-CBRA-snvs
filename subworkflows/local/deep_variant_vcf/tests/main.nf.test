nextflow_workflow {

    name "Test Subworkflow DEEP_VARIANT_VCF"
    script "../main.nf"
    workflow "DEEP_VARIANT_VCF"

    tag "subworkflows"
    tag "subworkflows/deep_variant_vcf"
    tag "subworkflows/mapping"
    tag "bwa"
    tag "bwa/mem"
    tag "picard"
    tag "picard/sortsam"
    tag "deepvariant"
    tag "bcftools"
    tag "bcftools/filter"



    setup {

        run("BWA_INDEX") {
            script "../../../../modules/nf-core/bwa/index/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        run("PICARD_CREATESEQUENCEDICTIONARY") {
            script "../../../../modules/nf-core/picard/createsequencedictionary/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        run("MAPPING") {
            script "../../../../subworkflows/local/mapping/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id : 'test_sample' ],
                    [file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)]
                ])
                input[1] =  Channel.fromList([
                            tuple([ id: 'test_sample'],
                            [])
                        ])
                input[2] = BWA_INDEX.out.index
                input[3] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    ])
                input[4] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[5] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz', checkIfExists: true)
                input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi', checkIfExists: true)
                """
            }
        }
    }


      test("deep_variant_vcf - bam") {
        when {
            workflow {
                """
                input[0] =  MAPPING.out.bam
                input[1] =  Channel.empty()
                input[2] =  Channel.fromList([
                    tuple([ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                ])
                input[3] = Channel.fromList([
                    tuple([ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}


