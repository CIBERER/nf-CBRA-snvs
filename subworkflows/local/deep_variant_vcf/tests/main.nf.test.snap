{
    "deep_variant_vcf - bam": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "vcf": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-27T15:57:11.845515547"
    }
}